docs: Updates to the Project Structure article (#1416)
NativeScript/docs: docs/core-concepts/project-structure.md

position: 120
slug: structure
previous_url: /structure
---

{% angular %}
{% endangular%} 

{% nativescript %}

> {% endnativescript %}

Reviewer: ?

Author: I am not sure why this one, the "TODOs" and the missing closing brackets are still present in this PR. They are not in my local repo and I am sure that I pushed. Anyway, thanks for pointing them out and sorry :))

# Project Structure

{% nativescript %}

> The default structure of a blank NativeScript project consist of a root folder that contains the `app`, `platforms` and `node_modules` directories, and a `package.json` configuration file. 

Reviewer: ```suggestion
The default structure of a blank NativeScript project consists of a root folder that contains the `app`, `platforms` and `node_modules` directories, and a `package.json` configuration file. 
```

```
myApplication/
└── app
    ├── App_Resources
    └── ...
├── node_modules
├── platforms
└── package.json
```
{% endnativescript %}

{% angular %}

> The default structure of a blank NativeScript + Angular project consist of a root folder that contains the `src`, `platforms`, `node_modules` and `hooks` directories, and several configuration files amongs which the most important is the `package.json`. 

Reviewer: ```suggestion
The default structure of a blank NativeScript + Angular project consists of a root folder that contains the `src`, `platforms`, `node_modules` and `hooks` directories, and several configuration files—amongst which the most important is the `package.json`. 
```

You can develop shared functionality or design in common files. To indicate that a file is common, make sure that the file name does not contain a `.android.` or `.ios.` string.

{% nativescript %}In the `app` folder, you will also find the `App_Resources` directory.{% endnativescript %}

### {% nativescript %}app{% endnativescript %}{% angular %}src{% endangular%}/package.json

This is a secondary `package.json` file in which you can specify the entry point file of the application and also several [Android-specific custom flags]({% slug custom-flags %}). Below is an example of a basic secondary `package.json` file.

% nativescript %}
```JSON
{
    "main": "app.js",
    "android": {
        "v8Flags": "--expose_gc",
        "forceLog": true,
        "discardUncaughtJsExceptions": true
    }
}
```
{% endnativescript %}
% angular %}
```JSON
{
    "main": "main.js",
    "android": {
        "v8Flags": "--expose_gc",
        "forceLog": true,
        "discardUncaughtJsExceptions": true
    }
}
```
{% endangular %}

##{% nativescript %}# app/{% endnativescript %}App_Resources

> The `App_Resources` folder contains the platform-specific resources of the application (icons, configuration files etc.). The configuration files that are respected by NativeScript tooling are respectively the `App_Resources/Android/src/main/AndroidManifest.xml` for Android and the `App_Resources/iOS/Info.plist` for iOS.

Reviewer: ```suggestion
The `App_Resources` folder contains the platform-specific resources of the application (icons, configuration files etc.). The configuration files that are respected by NativeScript tooling are the `App_Resources/Android/src/main/AndroidManifest.xml` for Android, and the `App_Resources/iOS/Info.plist` for iOS.
```

> (!!!!! FIX PATHS TO FILES !!!!)

Reviewer: :)

Additionally, you can modify the `App_Resources/iOS/build.xcconfig` or `App_Resources/Android/app.gradle` to add or remove additional build properties for the iOS and Android platforms, respectively.

## The **platforms** directory

> The `platforms` directory is created when you start a build or add a target platform to your project. The NativeScript tooling creates a new subdirectory with the respective platform name. These subdirectories have the platform-specific project structure required for native development with the native SDKs of the platform. When the project is prepared for build, the NativeScript tooling copies relevant content from the % nativescript %}`app`{% endnativescript %}{% angular %}`src`{% endangular%} directory to the platform-specific subdirectory for each target platform.

Reviewer: ```suggestion
The `platforms` directory is created when you start a build or add a target platform to your project. The NativeScript tooling creates a new subdirectory with the respective platform name. These subdirectories have the platform-specific project structure required for native development with the native SDKs of the platform. When the project is prepared for build, the NativeScript tooling copies relevant content from the {% nativescript %}`app`{% endnativescript %}{% angular %}`src`{% endangular%} directory to the platform-specific subdirectory for each target platform.
```

> > **IMPORTANT:** Avoid editing files located in the `platforms` subdirectories because the NativeScript CLI overrides them with the content of the % nativescript %}`app`{% endnativescript %}{% angular %}`src`{% endangular%} directory during the `prepare <Platform>` process.

Reviewer: ```suggestion
> **IMPORTANT:** Avoid editing files located in the `platforms` subdirectories because the NativeScript CLI overrides them with the content of the {% nativescript %}`app`{% endnativescript %}{% angular %}`src`{% endangular%} directory during the `prepare <Platform>` process.
```

## The **package.json** file

The main `package.json`, which is located in the root directory of the project, contains details about the application, its dependencies and other helpful information. You can set [common npm package.json properties](https://docs.npmjs.com/files/package.json) like `author`, `description` and `version`, or specify the npm packages and [NativeScript plugins]({% slug plugins-infrastructure %}) on which your app depends by modifying the `dependencies` and `devDependencies` attributes.

The root `package.json` also contains several NativeScript-specific properties which are placed inside the `nativescript` object:

> * **id** - Specifies the unique application identifier (App ID) of the app. To be able to build for both Android and iOS, your App ID must be unique and contain two or more strings, separated by a dot. Each string must start with a letter and should contain only letters and numbers. The app identifier must not start with an uppercase letter. (!!!!! LINK DOCUMENT ABOUT SEPARATE IDENTIFIERS -> NS 5.0 release)

Reviewer: Just fyi you have another placeholder in here.

* **tns-android.version** - Specifies the version of the Android runtime. If the property is missing, the latest version of the runtime will be added on the first run or build for Android.
* **tns-ios.version** - Specifies the version of the iOS runtime. If the property is missing, the latest version of the runtime will be added on the first run or build for iOS.

Here is an example of a basic main `package.json` file:

```JSON
{
    "nativescript": {
        "id": "org.nativescript.myApplication",
        "tns-android": {

>             "version": "4.2.0"

Reviewer: Probably worth worth making all of these 5.0.

        },
        "tns-ios": {
            "version": "4.2.0"
        }
    },
    "description": "My NativeScript Application",
    "license": "MIT",
    "repository": "https://github.com/myApplication",
    "dependencies": {
        "nativescript-theme-core": "~1.0.4",
        "tns-core-modules": "~4.2.0"
    },
    "devDependencies": {
        "nativescript-dev-typescript": "~0.7.0",
        "typescript": "~2.7.2"
    },
    "readme": "My NativeScript Application"
}
```

## The **hooks** directory

The `hooks` folder exists only when the project depends on plugins that require a hook to function properly. Hooks are executable pieces of code or Node.js scripts that are used to alter or augment the behavior of an extendable NativeScript CLI command. For more information about hooks and how to use them in NativeScript, see [Extending the CLI](https://github.com/NativeScript/nativescript-cli/blob/master/extending-cli.md).

Some of the more common plugins that have hooks are `nativescript-dev-webpack`, `nativescript-dev-typescript` and `nativescript-dev-sass`.

## The **tsconfig.json** file

> The `tsconfig.json` file is present only in projects that use TypeScript and its function is to work as a guide during the [transpilation]({% slug transpilers %}) of TypeScript to JavaScript. You can fine-tune the transpilation process by configuring the various [compiler options](https://www.typescriptlang.org/docs/handbook/compiler-options.html). For more information about `tsconfig.json`, see the official [TypeScript documentation](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html).

Reviewer: ```suggestion
The `tsconfig.json` file is present only in projects that use TypeScript. The file works as a guide during the [transpilation]({% slug transpilers %}) of TypeScript to JavaScript. You can fine-tune the transpilation process by configuring the various [compiler options](https://www.typescriptlang.org/docs/handbook/compiler-options.html). For more information about `tsconfig.json`, see the official [TypeScript documentation](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html).
```
